packages/history/index.ts

```diff
@@ -1,25 +1,94 @@
-const PopAction = 'POP';
-const PushAction = 'PUSH';
-const ReplaceAction = 'REPLACE';
+////////////////////////////////////////////////////////////////////////////////
+// TYPES
+////////////////////////////////////////////////////////////////////////////////
+
+export type Action = 'POP' | 'PUSH' | 'REPLACE';
+
+export type Path = string;
+export type PathPieces = {
+  pathname?: string;
+  search?: string;
+  hash?: string;
+};
+
+export type State = object;
```

Comment:
Would `unknown` be a better default? `object` excludes primitives such as `number`s, but there's no reason to restrict state to non-primitives?

Reply:
Actually:

> The state object can be anything that can be serialized. 

https://developer.mozilla.org/en-US/docs/Web/API/History/pushState#Parameters

So we should exclude functions amongst other things.

Reply:
Thanks for commenting here. I've updated this type to be `object | null`, but I'm not sure that's exactly what we want. Ideally I'd like this type to not be too restrictive. I just want to require some kind of an object, not a primitive, so `object` seemed like a good fit. Yes, there will be problems if it is not serializable, but I'm not aware of a nice, simple way to specify that in TS.